### DISCUSSION TITLE
Osu doesn't process click events from absolute pointer devices (pen tablet) on SDL's Wayland backend with Mutter.

### DISCUSSION THREAD

**Discussion by sp1ritCS** [ppy/osu#16851] [Q&A] [answered]
Since release 2022.205.0 apparently ships with SDL defaulting to Wayland I've noticed this. When setting SDL_VIDEODRIVER to x11 it has no issues and also setting SDL_VIDEODRIVER to wayland and launching rel. 2022.128.0 breaks it.

The input is provided by the compositor as opentabletdriver doesn't support it, but the kernel does. The pointer also moves as expected, only actually pressing down on the surface doesn't result in the click event one'd expects. This apparently also just affects GNOME Mutter as with wlroots/swaywm osu continues to process the click events flawlessly.

This also just affects pen input, input with mouse or even the touch surface of the same digitizer works.

However this is a very weird issue as doesn't appear to be an issue with SDLs input stack. I'm currently in the process of implementing an kernel (evdev) tablet handler for osu-framework since I believe this to be a better option to opentabletdriver as it'd allow for pseudo-exclusive access to the device when osu is running, without limiting usability when it's not (otd requires the wacom kernel module to be unloaded as it needs exclusive access to the device via hidraw). However there to I'm facing the issue of clicks (`MouseButtonInput` with `osuTK.Input.MouseButton.Left`) not being handled even tho they are clearly enqueued into osu with wayland, while x11 doesn't have any issues with it.

Any ideas how to debug this further?

**Comment by bdach** [comment, marked as answer]
sorry but you're going to have to figure this out on your own. we mostly just rely on SDL to work. if it doesn't work with some specific configuration then that's not really something any of us here is super interested in digging into because chances are it's not a problem with anything we're doing.

just to comment on one point, though:

> I'm currently in the process of implementing an kernel (evdev) tablet handler for osu-framework since I believe this to be a better option to opentabletdriver as it'd allow for pseudo-exclusive access to the device when osu is running, without limiting usability when it's not (otd requires the wacom kernel module to be unloaded as it needs exclusive access to the device via hidraw)

have you considered talking to the OTD devs about this at all in any capacity? going immediately to reimplementing the entire input stack using evdev seems like a radical reaction to me. work has already been done, there's no reason not to try cooperation if possible before going off on your own.

**Comment by sp1ritCS** [comment]
I apologize, I confused various issues.

I did some digging, and apparently Mutter exposes Tablet devices on Wayland via waylands `zwp_tablet_*v2` api (https://wayland.app/protocols/tablet-unstable-v2). While wlroots [seems to also support it](https://developer-docs.wacom.com/intuos-cintiq-business-tablets/docs/wayland-basics#compatibility) they also provide a "normal" wl_pointer for applications that do not make use of it. This didn't occur on XWayland, as XWayland consumes the output from the wl tablet api and handles it like a normal x cursor to keep comparability with applications written for xf86-input-wacom.

However libSDL2 doesn't support this tablet interface yet, but I didn't catch that, because I only tested against the libSDL2 from my distros repos (2.0.20), which didn't switch to default wayland (https://github.com/libsdl-org/SDL/pull/4306) yet.

I assume, since I doubt osu(-framework) would like to handle wayland specifics, I should try to get support of handleting `zwp_tablet_*v2` as pointer input device into libsdl2, in order to make it work in every software utilizing libsdl2?

---

But all of that has nothing to do with evdev tablet handler. I don't know why (will debug later), but the click event only appears to malfunction on the start screen (the one with the big "osu" button one has to press in order to get into the main menu) on wayland, after that input seems to work, I just failed to realize this this 🤦🏻‍♂️ .

In regards to
> have you considered talking to the OTD devs about this at all in any capacity?

I'm in the process of adding support for my Wacom DTH-1320 to OTD, however this is currently on hold from my side until I'm able to build/run a dotnet6 toolchain.

However the evdev handler is entirely separate from OTD, since the core idea of OTD (being cross-platform, by accessing the tablet via whatever interface the platform has to interact with raw usb streams and doing everything in userspace, thus requiring exclusive access to the device) is preventing input-wacom from exposing the device via evdev (and the other way around).

But maybe they aren't entirely separate, your right I could try to work with them in regards to reusing their area translation/transformation/rotation code, since as of now, I do translation and transformation calculations myself, but in a very unclean way. Maybe OTD's calculations are better (assuming they aren't performing worse).

**Reply by bdach** [reply]
> I assume, since I doubt osu(-framework) would like to handle wayland specifics, I should try to get support of handleting `zwp_tablet_*v2` as pointer input device into libsdl2, in order to make it work in every software utilizing libsdl2?

adding support upstream would be preferable yes. however the tablet API you mention seems to be marked "unstable", so i imagine upstream may want to hold off on that until it stabilises.

**Reply by sp1ritCS** [reply]
> however the tablet API you mention seems to be marked "unstable", so i imagine upstream may want to hold off on that until it stabilises.

there are quite a few wayland protocols still marked unstable, sdl also seems to make use of them (https://github.com/libsdl-org/SDL/tree/main/wayland-protocols). I've created an issue: https://github.com/libsdl-org/SDL/issues/5344. Lets see how it goes :)

**Reply by smoogipoo** [reply]
Just going to say - I don't think we want to go down the path of evdev handler in any case. It means reimplementing everything OTD ourselves, which means that there's going to be tablet-specific edge cases (which we've previously had to deal with in raw input handling) and on top of that we can't make use of other OTD niceties such as anti-chatter + smoothing which people have requested us to add. It's an entire input stack that is too much to handle for us.

Best would be to make libSDL work for your setup and call it a day at that.

**Reply by sp1ritCS** [reply]
> It means reimplementing everything OTD ourselves, which means that there's going to be tablet-specific edge cases (which we've previously had to deal with in raw input handling)

yeah. That's something I've noticed too. I'm considering to go one level higher by using [libinput](https://www.freedesktop.org/wiki/Software/libinput/). It has such niceties as not having to do translation/transformation/rotation math oneself but instead takes a [calibration matrix](https://wayland.freedesktop.org/libinput/doc/latest/absolute-axes.html#calibration-of-absolute-devices), it having access to the systemd/udev hwdb also means it can provides the position in metric units (mm) instead of the raw sensor units evdev provides and the idea behind is to [unify device specific quirks](https://lwn.net/Articles/760770/).

However I'd need to link against libinput and require it to keep it's symbols unlike the kernel where I can just use evdev ioctls and their "never break userspace" guarantee, so I have no idea if you'd consider it mergable for osu-framework .

---

The SDL patch I've posted works, however it still goes over the compositor. With my current implementation of evdev support in osu-framework, I believe (might be placebo effect tho) to have lower latency than via SDL since it doesn't have to go over X and/or the Wayland compositor.

Should "wiring up the whole thing could be done as a bigger feature later" mean, that SDL will provide an actual Pen-Tablet interface at some point, it could use evdev/libinput directly if it has read access to the device, but fall back to the `zwp_tablet_*v2` api when it hasn't, like it does already for pointing devices.
